platform/machine/esx/cluster.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package esx
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/satori/go.uuid"
+
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/api/esx"
+	"github.com/coreos/mantle/platform/conf"
+)
+
+var (
+	plog = capnslog.NewPackageLogger("github.com/coreos/mantle", "platform/machine/esx")
+)
+
+type cluster struct {
+	*platform.BaseCluster
+	api *esx.API
+}
+
+// NewCluster creates an instance of a Cluster suitable for spawning
+// instances on Amazon Web Services' Elastic Compute platform.
+//
+// NewCluster will consume the environment variables $ESX_HOST, $ESX_USER,
+// and $ESX_PASSWORD to spawn instances and authenticate.
+func NewCluster(opts *esx.Options, rconf *platform.RuntimeConfig) (platform.Cluster, error) {
+	api, err := esx.New(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	bc, err := platform.NewBaseCluster(opts.BaseName, rconf)
+	if err != nil {
+		return nil, err
+	}
+
+	ec := &cluster{
+		BaseCluster: bc,
+		api:         api,
+	}
+
+	return ec, nil
+}
+
+func (ec *cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
+	conf, err := ec.RenderUserData(userdata, map[string]string{
+		"$private_ipv4": "${COREOS_ESX_IPV4_PRIVATE_0}",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	conf.AddSystemdUnit("coreos-metadata.service", "[Unit]\nDescription=VMware metadata agent\n\n[Service]\nType=oneshot\nEnvironment=OUTPUT=/run/metadata/coreos\nExecStart=/usr/bin/mkdir --parent /run/metadata\nExecStart=/usr/bin/bash -c 'echo \"COREOS_ESX_IPV4_PRIVATE_0=$(ip addr show ens192 | grep -Po \"inet \\K[\\d.]+\")\" > ${OUTPUT}'\n", false)
+
+	id := uuid.NewV4()
+
+	instance, err := ec.api.CreateDevice(id.String(), conf)
```

Comment:
The device name should be `ec.Name()` plus a random suffix.  That suffix can use some random bytes as other drivers do; a full UUID is unnecessary.